codalab/model/bundle_model.py

```diff
@@ -677,6 +677,85 @@ def get_bundle_uuids(self, conditions, max_results):
 
         return self._execute_query(query)
 
+    def get_memoized_bundles(self, user_id, command, dependencies):
+        '''
+        Get the most recently finished bundle that matches with input command and dependencies.
+        :param user_id: a string that specifies the current user id.
+        :param command: a string that defines the command that is used to search for memoized bundles in the database.
+        :param dependencies: a list of string in the structure of ["key1:uuid1","key2:uuid2"] defines
+                            dependencies that are used to search for memoized bundles in the database.
+        :return: a list of matched uuids.
+        '''
+        # Get a list of uuids that matches with the given dependencies from the cl_bundle table.
+        # This step is to standardize uuid for future comparison.
+        with self.engine.begin() as connection:
+            # Select bundles that has the given command from the bundle table.
+            filter_on_command = (
+                select([cl_bundle.c.uuid])
+                .distinct()
+                .select_from(cl_bundle)
+                .where(and_(cl_bundle.c.command == command, cl_bundle.c.owner_id == user_id))
+                # Ensure the order of the returning bundles will be in the order of they were created.
+                .order_by(cl_bundle.c.id)
+            )
+            rows = connection.execute(filter_on_command).fetchall()
+            if not rows:
+                return []
+            uuids = [row[0] for row in rows]
+            # When there is no dependency to be matched, the target memozied bundle
+            # should only exist in the bundle table but not in the bundle_dependency table.
+            if len(dependencies) == 0:
+                query = (
+                    select([cl_bundle_dependency.c.child_uuid])
+                    .distinct()
+                    .select_from(cl_bundle_dependency)
+                    .where(cl_bundle_dependency.c.child_uuid.in_(uuids))
+                )
+                dependency_rows = connection.execute(query).fetchall()
+                dependency_uuids = set([row[0] for row in dependency_rows])
+                return [uuid for uuid in uuids if uuid not in dependency_uuids]
+
+            dep_clause = []
+            for dep in dependencies:
+                key, uuid = dep.split(':')
```

Comment:
bad to have raw manipulation, should use standard functions

Reply:
I have been trying to get rid of this manipulation coz I do think it's not a good practice here. But the intention behind this is primarily trying to be able to search for a list of `(key, uuid)` pair that can match with the input command + dependencies in the database. If we take a look in the schema we have, `cl_bundle_dependency` has the following columns:
|index| column | 
|-- |-- |
| 1. | child_uuid | 
|2. | child_path |
|3.| parent_uuid|
|4.| parent_path|

`child_uuid` is the target bundle that we will be searching for. `key` will map to `child_path` and `uuid` will map to `parent_uuid`. Given that the relationship between `child_uuid` and `child_path (key), parent_uuid (uuid)` is `1-to-n` mapping, when we group by `child_uuid`, we need some ways to keep the records in the other two columns together. I have been searching for solutions to keep those records without doing any concatenation, but so far I haven't found anything yet. If you guys @teetone or @percyliang have any suggestions, please let me know.

Reply:
I think I understand the situation. When the `/bundles` endpoint is called, can we have the contract to be that a JSON list of `child_path` -`parent_uuid`  pairs has to be passed as the dependencies? Something like this:
```
x = '[{ "child_path":"some child", "parent_uuid": "0xblah"}, 
      { "child_path":"some child2", "parent_uuid": "0xblah2"}]'
```
Then, we can parse the JSON on the server-side using `json.loads`, so `json.loads(x)` should give you the list of dependencies you need. The other alternative (but I think less cleaner) solution would be to just pass in two separate lists.

Reply:
I think this is a great idea! Let me change this.